Feat/golds

esnyder232 opened the pull request:
Got golds working.
Golds are calculated on the backend with a function call on chat.js
Added 2 coolhole modules:
 - coolholecommon - sets up common objects for other coolhole modules.
 - coolholegolds - contains all the gold calculation logic

esnyder232 committed:
backend done for now

esnyder232 committed:
got golds mostly working

esnyder232 committed:
coolhole-server-mod hooks

esnyder232 committed:
fixed some css bugs and my own mistakes

esnyder232 committed:
merging 3.0

esnyder232 committed:
minor changes

streeeggs commented on src/channel/coolhole-server-mods.js:
```diff
@@ -0,0 +1,31 @@
+        return true;
+    }
+
+    var allow = true;
```
nit: personally, I avoid `var` unless globally scoped. 
```suggestion
    let allow = true;
```

streeeggs commented on src/channel/chat.js:
```diff
@@ -335,6 +336,9 @@ ChatModule.prototype.processChatMsg = function (user, data) {
     }
 
+    if(!coolholePostProcessChatMessage(this.channel, user, data, msgobj))
+        return;
```
question: do we really care what `coolholePostProcessChatMessage` returns? And do we want to end this method early if something goes tits up? 

If our intention is to have regular cytube functionality "just werk" if our code fails, what our methods return probably doesn't matter as long as we don't mutate arguments

streeeggs replied:
Speaking of mutating arguments, I generally think that sucks haha. I know sync does that everywhere but IMO just returning a new object to merge in to the `msgobj` seems must easier to handle.

esnyder232 replied:
> question: do we really care what `coolholePostProcessChatMessage` returns? And do we want to end this method early if something goes tits up?
> 
> If our intention is to have regular cytube functionality "just werk" if our code fails, what our methods return probably doesn't matter as long as we don't mutate arguments

I coded it this way because of other coolhole features that eventually need to be created. It is to provide a way of "denying" the chat message from going through, its mainly for stuff like the cp points allow/deny the chat message if they have enough points (like /secretary, /danmu, /highlight). I could have just called "coolholePostProcessChatMessage()" without the if check and return, but i was just looking ahead i guess.


> Speaking of mutating arguments, I generally think that sucks haha. I know sync does that everywhere but IMO just returning a new object to merge in to the `msgobj` seems must easier to handle.

Well, i was returning "true/false" for the allow/deny functionality. I was also trying to minimize the # of code changes made to cytube files. If we just do something like "msgobj = coolholePostProcessChatMessage(...)", its essentially accomplishing the same thing as mutating the arguments itself. Unless we want to do something like:
`
results = coolholePostProcessChatMessage(this.channel, user, data, msgobj);
if(results.allow)
    msgobj = results.msgobj
else
    return;
`
I was just trying minimize the logic/code on the cytube files thats all.

esnyder232 replied:
Okay, my mistake. I looked at the code again for current coolhole. You can disregard my last comment about the "denying the chat message".

In current coolhole prod, all the features that deny the chat message (/secretary, /danmu, /highlight) are actually handled in the coolpoints module by simply returning early and not calling "processChatMessage()" again.

So, my code is redundant, my bad. I think I was trying to cram too much in my brain while doing it and I thought "oh yeah, we'll definitely need this for later" and just got messed up. I'll remove the if statement.

As for the "mutating argument", the handleDebtEffects() also needs to be called modify the text, and we basically modified the msgobj directly in coolhole prod. Though we could do a similar thing there, just do a `msgobj.msg = coolholeHandleDebtEffects(...);" or something like that.

Maybe you're right, and maybe I'm trying to do too much. I'll change it so only my "coolholeMeta" object is attached to msgobj in chat and see how that goes.

cck37 replied:
>It is to provide a way of "denying" the chat message from going through, its mainly for stuff like the cp points allow/deny the chat message if they have enough points (like /secretary, /danmu, /highlight)

Oh I see. We don't want to emit a chatmsg event if they have insufficient funds. Then yea I think you're right. We'd probably wanna do something like you suggested
```js
try {
  const coolholeMeta = coolholePostProcessChatMessage(this.channel, user, data, msgobj);
  if (coolholeMeta.allow)
    msgObj.meta.coolholeMeta = coolholeMeta;
  else
    // either return or do our socket emit logic here... probably better to stuff this server-mod
    user.socket.emit("coolholeDisallow", {
      action: coolholeMeta.error.action,
      message: coolholeMeta.error.message
    }
 }
 catch (Execption e) {
   // Probably wanna log more than just exception 
   console.error("Some shit happened in coolhole", e);
   this.channel.logger.log("Some shit happened in coolhole" + JSON.stringify(e));
   // Ignoring uncaught exception
}
```


>In current coolhole prod, all the features that deny the chat message (/secretary, /danmu, /highlight) are actually handled in the coolpoints module by simply returning early and not calling "processChatMessage()" again.

Yeah I think that _sucks_ haha. It was a naive approach that just assumed nothing would ever fail. I'm trying to advocate for something like a try catch around our code in `chat.js` and just make sure we leave the `msgObj` unmolested if it blows up. This way, regular chat functionality should just work if our code blows up. IE: I'd rather see `/danmu hey` in the chat rather than nothing appearing with no feedback.

>I'll change it so only my "coolholeMeta" object is attached to msgobj in chat and see how that goes.

Yeah this is really what I was trying to say. This way we don't, say, mutate the object partially, have coolhole code fail, and then the message object be partially updated.
It's also generally pretty painful to think about how the object was mutated in a function call instead of just thinking what a function returns.

streeeggs replied:
>It is to provide a way of "denying" the chat message from going through, its mainly for stuff like the cp points allow/deny the chat message if they have enough points (like /secretary, /danmu, /highlight)

Oh I see. We don't want to emit a chatmsg event if they have insufficient funds. Then yea I think you're right. We'd probably wanna do something like you suggested
```js
try {
  const coolholeMeta = coolholePostProcessChatMessage(this.channel, user, data, msgobj);
  if (coolholeMeta.allow)
    msgObj.meta.coolholeMeta = coolholeMeta;
  else
    // break the chat message flow; CH should emit its own events and only its own events
    return;
    /* probably better to stuff this server-mod
    user.socket.emit("coolholeDisallow", {
      action: coolholeMeta.error.action,
      message: coolholeMeta.error.message
    }
    */
 }
 catch (Execption e) {
   // Probably wanna log more than just exception 
   console.error("Some shit happened in coolhole", e);
   this.channel.logger.log("Some shit happened in coolhole" + JSON.stringify(e));
   // Ignoring uncaught exception
}
```

>In current coolhole prod, all the features that deny the chat message (/secretary, /danmu, /highlight) are actually handled in the coolpoints module by simply returning early and not calling "processChatMessage()" again.

Yeah that was "smart" of me I guess since we have to override what the server would emit when sending a message. If we wanted to avoid modifying the chatMsg client side code entirely, we'd need a way to fork from the usual sockets called just below this. I think what you were suggesting is probably the best approach.

>I'll change it so only my "coolholeMeta" object is attached to msgobj in chat and see how that goes.

Yeah this is really what I was trying to say. This way we don't, say, mutate the object partially, have coolhole code fail, and then the message object be partially updated. IE: I'd rather see `/danmu hey` in the chat rather than nothing appearing with no feedback. 

It's also generally pretty painful to think about how the object was mutated in a function call instead of just thinking what a function returns.

streeeggs commented on src/channel/chat.js:
```diff
@@ -5,6 +5,7 @@ var util = require("../utilities");
 var Flags = require("../flags");
 import { transformImgTags } from '../camo';
 import { Counter } from 'prom-client';
+import { coolholePostProcessChatMessage } from './coolhole-server-mods';
```
interesting. you _didn't_ want to type `this.channel.modules` 😏

streeeggs commented on src/channel/coolholecommon.js:
```diff
@@ -0,0 +1,27 @@
+var ChannelModule = require("./module");
```
nit: `import` instead of require if possible

streeeggs commented on src/channel/coolholegolds.js:
```diff
@@ -0,0 +1,76 @@
+ * @returns always return true to let the chat message passthrough
+ */
+CoolholeGoldsModule.prototype.calculateGold = function (msgobj) {
+    if(this.channel.modules["playlist"].current === null ||
+       this.channel.modules["playlist"].current.media === null ||
+       this.channel.modules["playlist"].current.media.title === null)
```
```suggestion
    if(this.channel.modules["playlist"]?.current?.media?.title === null)
```

streeeggs commented on src/channel/coolholegolds.js:
```diff
@@ -0,0 +1,76 @@
+    // 1% chance? Idk either. If it's equal to one, they get a gold
+    if (lotteryHash === 1) {
+        LOGGER.info(`calculateGold: Found Gold Message: ${chatText}`);
+        msgobj.meta.coolholeMeta.otherClasses.push("text-lottery");
```
question: do we wanna add european gold here too or leave that in js/css for now?

esnyder232 replied:
Huh, I didn't think about that actually. Maybe it would be cleaner on the server side?

We can probably just leave it the way it is for now...there are other features like points, sounds, options that should probably be completed before we refactor a bit.

cck37 replied:
Yeah I agree. Let's work towards an MVP and get something close to how CH works currently (custom JS and all). Then we can start porting over mainstay features into source

streeeggs replied:
Yeah I agree. Let's work towards an MVP and get something close to how CH works currently (custom JS and all). Then we can start porting over mainstay features into source

streeeggs commented on www/js/coolhole-util.js:
```diff
@@ -0,0 +1,36 @@
+
+    }
+    catch(ex) {
+        console.log(ex);
```
nit: should be console.error IMO

streeeggs commented:
All in all, looks good to me. Only real concerns are the placement of our chat mod function and the mutating of props (ie: `msgObj`). 
- There's a small worry that we might want to apply certain functions before or after in the sync tube flow (replacing for filters, emotes, etc). I can't think of a better way than what you presented so maybe just something to be aware of for now...
- Personally in favor of avoiding that and only blending our object with the sync object once everything is completed. This way, if things go tits up, we don't have to worry about some partial chat object that has like the `danmu` class but the user never actually paid for it.

esnyder232 commented:
Thanks for looking at it streeegs btw.

>     * There's a small worry that we might want to apply certain functions before or after in the sync tube flow (replacing for filters, emotes, etc). I can't think of a better way than what you presented so maybe just something to be aware of for now...

Yeah, I agree there might be others. Hard to tell.
Currently in coolhole prod, we check for golds and handle debt effects inside "ChatModule.prototype.formatMessage". This function actually gets called twice when there is a command as well (like /say or /danmu), so its not really a good spot for it even though the function name "formatMessage" is appropriate. I tried various other places to put the logic, but resulted in more code in cytube files (onUserPreChat hook, and try creating my own hook). So I settled on just a "coolholePostXYZ()" style of doing it, where XYZ is the function being called, and put the function call near the end.

If we do we need other functionality before replacing filters, emotes, etc, we can do a similar thing and do "coolholePreXYZ()" and just place the function call in the beginning of the function (like "coolholePreProcessChatMessage(...)" or something). Other than that, I can't think of another way to do it with minimal changes.

streeeggs commented:
> If we do we need other functionality before replacing filters, emotes, etc, we can do a similar thing and do "coolholePreXYZ()" and just place the function call in the beginning of the function (like "coolholePreProcessChatMessage(...)" or something)

Yeah I think that's fair for now. It would be nice if we could easily hook into the chat event before and after this does without impacting much of cytube but I think this is a fair compromise for now

esnyder232 committed:
applied changes from pr. Also moved 'coolhole-server-mods' function to coolholecommon module.

esnyder232 committed:
replaced requires with imports

esnyder232 commented:
I removed the "coolholer-server-mod.js" file and just moved that function to coolholecommon module.
It just seemed kind of wierd to have it at that point...might as well just shove the logic in the module.
Also had to add the "otherClasses" to the channel logging (I forgot when i first did the PR).
